## NLP TWITTER ANALYSIS

* Final Project Submission
* Group Members
   1. Benson Kamau
   2. Kevin Muchori
   3. Nancy Chelangat
   4. Sally Kinyanjui
   5. Breden Mugambi

* Student Pace: Full-Time
* Instructor's: Nikita Njoroge

### Problem Statement
Accurately classifying the sentiments expressed in tweets about topics or brands into specific classes- positive, negative or neutral is a huge challenge for companies like Apple and Google. Given the diverse nature of informal data, with its use of slang, abbreviations, coming up with a reliable sentiment analysis model that can effectively interpret and classify the tweets can be a complex task. Getting this task right provides a wide variety of novel information for a company like Apple by providing insights and creating better understanding overall of how consumers interact with products/brands.

### Objective
The main objective is to build a model that can rate the sentiment of  a tweet based on its content.

#### Project Success Metrics
Over 75% accuracy on the testing data.

#### Importing relavent Libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import nltk
from nltk.tokenize import regexp_tokenize, word_tokenize,TweetTokenizer, RegexpTokenizer
from nltk.corpus import stopwords, wordnet
from nltk.stem import PorterStemmer
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
import re
import nltk
nltk.download('punkt')



[nltk_data] Downloading package punkt to /Users/mac/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

### 1. Data Loading and Understanding

The dataset that will be used in this study comes from CrowdFlower via data.world through this link - https://data.world/crowdflower/brands-and-product-emotions.


In [2]:
#create a function that loads data and gets the info about the data.
def load_and_get_info(file_path, encoding='utf-8'):
    try :
        # Load data
        df = pd.read_csv(file_path, encoding=encoding)

        # Display the first few rows of the DataFrame
        df_head = df.head()

        # Get information about the DataFrame
        df_info = df.info()

        return df,df_info, df_head
    except UnicodeDecodeError:
        print(f"Failed to decode {file_path} with encoding {encoding}. Trying with 'latin1' encoding.")
        return load_and_get_info(file_path, encoding='latin1')

# A function that checks the data types of DataFrame columns and return the count of columns for each data type category.
def check_data_types(df):

    data_type_counts = df.dtypes.replace({'object': 'string'}).value_counts().to_dict()
    return data_type_counts

In [3]:
file_path1 = 'tweet-analysis.csv'
df1,data_info, data_head = load_and_get_info(file_path1)
print(data_info)
print("\nFirst few rows of the DataFrame:")
data_head #data_head

Failed to decode tweet-analysis.csv with encoding utf-8. Trying with 'latin1' encoding.
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9093 entries, 0 to 9092
Data columns (total 3 columns):
 #   Column                                              Non-Null Count  Dtype 
---  ------                                              --------------  ----- 
 0   tweet_text                                          9092 non-null   object
 1   emotion_in_tweet_is_directed_at                     3291 non-null   object
 2   is_there_an_emotion_directed_at_a_brand_or_product  9093 non-null   object
dtypes: object(3)
memory usage: 213.2+ KB
None

First few rows of the DataFrame:


tweet_text  \
0  .@wesley83 I have a 3G iPhone. After 3 hrs twe...   
1  @jessedee Know about @fludapp ? Awesome iPad/i...   
2  @swonderlin Can not wait for #iPad 2 also. The...   
3  @sxsw I hope this year's festival isn't as cra...   
4  @sxtxstate great stuff on Fri #SXSW: Marissa M...   

  emotion_in_tweet_is_directed_at  \
0                          iPhone   
1              iPad or iPhone App   
2                            iPad   
3              iPad or iPhone App   
4                          Google   

  is_there_an_emotion_directed_at_a_brand_or_product  
0                                   Negative emotion  
1                                   Positive emotion  
2                                   Positive emotion  
3                                   Negative emotion  
4                                   Positive emotion

The dataset contains the following columns:

1. 'tweet_text' column
    - Contains the text of the tweet.
2. 'emotion_in_tweet_is_directed_at' column
    - Contains the person or entity that the tweet is directed at.
3. 'is_there_an_emotion_directed_at_a_brand_or_product' column
    - Indicates the kind of emotion in the tweet directed at the brand or product

The dataset has a total of 9093 data points.



In [4]:
#check the data types of DataFrame columns in our training set values.
data_type_counts = check_data_types(df1)
print("Count of columns for each data type category:")
print(data_type_counts)

Count of columns for each data type category:
{'string': 3}


The dataset has one data type category .i., object type.


To simplify working with the dataset, we will rename the columns to simpler and shorter names.


In [5]:
# function to rename the column names
def rename_columns(df, columns_dict):
    """
   Parameters:
    df (pd.DataFrame): The DataFrame whose columns need to be renamed.
    columns_dict (dict): A dictionary where keys are current column names and values are the new column names.
    """
    df.rename(columns=columns_dict, inplace=True)
    return df

# Define the dictionary for renaming columns
columns_dict = {
    'tweet_text': 'tweet',
    'emotion_in_tweet_is_directed_at': 'target_entity',
    'is_there_an_emotion_directed_at_a_brand_or_product': 'emotion'
}

# Rename columns using the dictionary
df1 = rename_columns(df1, columns_dict)

print("\nRenamed DataFrame columns:")
print(df1.columns)
df1.head()


Renamed DataFrame columns:
Index(['tweet', 'target_entity', 'emotion'], dtype='object')


tweet       target_entity  \
0  .@wesley83 I have a 3G iPhone. After 3 hrs twe...              iPhone   
1  @jessedee Know about @fludapp ? Awesome iPad/i...  iPad or iPhone App   
2  @swonderlin Can not wait for #iPad 2 also. The...                iPad   
3  @sxsw I hope this year's festival isn't as cra...  iPad or iPhone App   
4  @sxtxstate great stuff on Fri #SXSW: Marissa M...              Google   

            emotion  
0  Negative emotion  
1  Positive emotion  
2  Positive emotion  
3  Negative emotion  
4  Positive emotion

The dataset has been successfully renamed and the column names are now more descriptive.

# Data Cleaning
This is an essential aspect so as to ensure that the text data is consistent and free of errors. For this project, we will check for missing values, checking for duplicates,
remove white spaces, handle capitalization.



In [6]:
# Check missing values
df1.isna().sum()

tweet               1
target_entity    5802
emotion             0
dtype: int64

There is only one missing tweet text value, which will be removed. There are 5,802 missing "target_entity" values; however, this is acceptable since the current project focuses on overall tweet sentiment rather than specific items. These missing values will be replaced with an "Uncategorized" classification.

In [7]:
#Removing Null Tweets, Removing Duplicate entries and Filling in missing Item Values

#Removing 1 null 'Tweet' Entry
df1.dropna(subset = ['tweet'], inplace=True)

#Removing Duplicates
df1.drop_duplicates(inplace=True)

#Filling in Null "Item" categories with "Uncategorized"
df1['target_entity'].fillna('Uncategorized', inplace=True)

df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9070 entries, 0 to 9092
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   tweet          9070 non-null   object
 1   target_entity  9070 non-null   object
 2   emotion        9070 non-null   object
dtypes: object(3)
memory usage: 283.4+ KB


At this point there should not be any duplicate entries or null values in the data and the total row count in the dataset has decreased from 9092 to 9070.

Text Cleaning: The function below cleans the tweet column by removing white spaces, converting to lower case, and removing special characters.

In [8]:
def clean_text(text):
    if not isinstance(text, str):
        return ""
    text = text.strip()
    text = text.lower()
    pattern = re.compile(r'[^a-zA-Z0-9\s]')
    text = pattern.sub('', text)
    return text

# Apply the clean_text function to the tweet column
df1['cleaned_tweet'] = df1['tweet'].apply(clean_text)
df1.head()

tweet       target_entity  \
0  .@wesley83 I have a 3G iPhone. After 3 hrs twe...              iPhone   
1  @jessedee Know about @fludapp ? Awesome iPad/i...  iPad or iPhone App   
2  @swonderlin Can not wait for #iPad 2 also. The...                iPad   
3  @sxsw I hope this year's festival isn't as cra...  iPad or iPhone App   
4  @sxtxstate great stuff on Fri #SXSW: Marissa M...              Google   

            emotion                                      cleaned_tweet  
0  Negative emotion  wesley83 i have a 3g iphone after 3 hrs tweeti...  
1  Positive emotion  jessedee know about fludapp  awesome ipadiphon...  
2  Positive emotion  swonderlin can not wait for ipad 2 also they s...  
3  Negative emotion  sxsw i hope this years festival isnt as crashy...  
4  Positive emotion  sxtxstate great stuff on fri sxsw marissa maye...

The additional column cleaned_tweet stores the text after text cleaning.

## Lemmatization ,stop word removal and tokenization

In [9]:
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize

# Initialize the lemmatizer
lemmatizer = WordNetLemmatizer()

# function to lemmatize text
def lemmatize_text(text):
    tokens = word_tokenize(text)
    lemmatized_tokens = [lemmatizer.lemmatize(token) for token in tokens]
    return ' '.join(lemmatized_tokens)

# Apply the lemmatize_text function to the 'cleaned_tweet' column
df1['lemmatized_tweet'] = df1['cleaned_tweet'].apply(lemmatize_text)

df1.head(10)




tweet       target_entity  \
0   .@wesley83 I have a 3G iPhone. After 3 hrs twe...              iPhone   
1   @jessedee Know about @fludapp ? Awesome iPad/i...  iPad or iPhone App   
2   @swonderlin Can not wait for #iPad 2 also. The...                iPad   
3   @sxsw I hope this year's festival isn't as cra...  iPad or iPhone App   
4   @sxtxstate great stuff on Fri #SXSW: Marissa M...              Google   
5   @teachntech00 New iPad Apps For #SpeechTherapy...       Uncategorized   
7   #SXSW is just starting, #CTIA is around the co...             Android   
8   Beautifully smart and simple idea RT @madebyma...  iPad or iPhone App   
9   Counting down the days to #sxsw plus strong Ca...               Apple   
10  Excited to meet the @samsungmobileus at #sxsw ...             Android   

                               emotion  \
0                     Negative emotion   
1                     Positive emotion   
2                     Positive emotion   
3                     Negative emotion   
4                     Positive emotion   
5   No emotion toward brand or product   
7                     Positive emotion   
8                     Positive emotion   
9                     Positive emotion   
10                    Positive emotion   

                                        cleaned_tweet  \
0   wesley83 i have a 3g iphone after 3 hrs tweeti...   
1   jessedee know about fludapp  awesome ipadiphon...   
2   swonderlin can not wait for ipad 2 also they s...   
3   sxsw i hope this years festival isnt as crashy...   
4   sxtxstate great stuff on fri sxsw marissa maye...   
5   teachntech00 new ipad apps for speechtherapy a...   
7   sxsw is just starting ctia is around the corne...   
8   beautifully smart and simple idea rt madebyman...   
9   counting down the days to sxsw plus strong can...   
10  excited to meet the samsungmobileus at sxsw so...   

                                     lemmatized_tweet  
0   wesley83 i have a 3g iphone after 3 hr tweetin...  
1   jessedee know about fludapp awesome ipadiphone...  
2   swonderlin can not wait for ipad 2 also they s...  
3   sxsw i hope this year festival isnt a crashy a...  
4   sxtxstate great stuff on fri sxsw marissa maye...  
5   teachntech00 new ipad apps for speechtherapy a...  
7   sxsw is just starting ctia is around the corne...  
8   beautifully smart and simple idea rt madebyman...  
9   counting down the day to sxsw plus strong cana...  
10  excited to meet the samsungmobileus at sxsw so...

WordNetLemmatizer() is used to reduce words to their base or root form. This is what is known as lemmatization.

This function def lemmatize_text(text)  tokenizes the input text into words,lemmatizes each word and joins the lemmatized words back into a string.

From the output , the original tweet is the raw tweet text,the cleaned  tweet is the twwet after preprocessing  and the lemmatized tweet is the cleaned twwet after lemmatization.

In [10]:
# Load the stop words
stop_words = set(stopwords.words('english'))
stop_words

{'a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 'her',
 'here',
 'hers',
 'herself',
 'him',
 'himself',
 'his',
 'how',
 'i',
 'if',
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it's",
 'its',
 'itself',
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'only',
 'or',
 'other',
 'our',
 'ours',
 'ourselves',
 'out',
 'over',
 'own',
 'r

In the input, english indicates that the output required is supposed to be a setg of English stop words.

This set of stop words can be used in text preprocessing tasks

In [11]:
def preprocess_text_nltk(text):
    # Tokenize the text
    tokens = word_tokenize(text)
    
    # Remove stop words
    r_stop_words = [word for word in tokens if word.lower() not in stop_words]
  

# Initialize the lemmatizer
lemmatizer = WordNetLemmatizer()

# function to lemmatize text
def lemmatize_text(text):
    tokens = word_tokenize(text)
    lemmatized_tokens = [lemmatizer.lemmatize(token) for token in tokens]
    return ' '.join(lemmatized_tokens)

# Apply the lemmatize_text function to the 'cleaned_tweet' column
df1['lemmatized_tweet'] = df1['cleaned_tweet'].apply(lemmatize_text)
df1.head(10)

tweet       target_entity  \
0   .@wesley83 I have a 3G iPhone. After 3 hrs twe...              iPhone   
1   @jessedee Know about @fludapp ? Awesome iPad/i...  iPad or iPhone App   
2   @swonderlin Can not wait for #iPad 2 also. The...                iPad   
3   @sxsw I hope this year's festival isn't as cra...  iPad or iPhone App   
4   @sxtxstate great stuff on Fri #SXSW: Marissa M...              Google   
5   @teachntech00 New iPad Apps For #SpeechTherapy...       Uncategorized   
7   #SXSW is just starting, #CTIA is around the co...             Android   
8   Beautifully smart and simple idea RT @madebyma...  iPad or iPhone App   
9   Counting down the days to #sxsw plus strong Ca...               Apple   
10  Excited to meet the @samsungmobileus at #sxsw ...             Android   

                               emotion  \
0                     Negative emotion   
1                     Positive emotion   
2                     Positive emotion   
3                     Negative emotion   
4                     Positive emotion   
5   No emotion toward brand or product   
7                     Positive emotion   
8                     Positive emotion   
9                     Positive emotion   
10                    Positive emotion   

                                        cleaned_tweet  \
0   wesley83 i have a 3g iphone after 3 hrs tweeti...   
1   jessedee know about fludapp  awesome ipadiphon...   
2   swonderlin can not wait for ipad 2 also they s...   
3   sxsw i hope this years festival isnt as crashy...   
4   sxtxstate great stuff on fri sxsw marissa maye...   
5   teachntech00 new ipad apps for speechtherapy a...   
7   sxsw is just starting ctia is around the corne...   
8   beautifully smart and simple idea rt madebyman...   
9   counting down the days to sxsw plus strong can...   
10  excited to meet the samsungmobileus at sxsw so...   

                                     lemmatized_tweet  
0   wesley83 i have a 3g iphone after 3 hr tweetin...  
1   jessedee know about fludapp awesome ipadiphone...  
2   swonderlin can not wait for ipad 2 also they s...  
3   sxsw i hope this year festival isnt a crashy a...  
4   sxtxstate great stuff on fri sxsw marissa maye...  
5   teachntech00 new ipad apps for speechtherapy a...  
7   sxsw is just starting ctia is around the corne...  
8   beautifully smart and simple idea rt madebyman...  
9   counting down the day to sxsw plus strong cana...  
10  excited to meet the samsungmobileus at sxsw so...

Similar to lemmatization,both functions play the same role .

The function 'def preprocess_text_nltk(text)' is used to tokenize the text.

In [12]:
# Function to tokenize and remove stopwords and punctuation
def process_tweet(tweet):
    tweet = tweet.lower() #convert tweet to lowercase
   # for 
    pattern = r"\b\w+(?:'\w+)?\b"
    # Create a RegexpTokenizer with the defined pattern
    tokenizer = RegexpTokenizer(pattern)
    # Tokenize tweet
    tokens = tokenizer.tokenize(tweet)
    tokens = [token for token in tokens if token not in stopwords.words('english') 
              and token not in punctuation]                                   
    return tokens


Pattern r"\b\w+(?:'\w+)?\b"  used for tokenizing text while keeping words with apostrophes. Below is an example.

In [13]:
import re
from nltk.tokenize import RegexpTokenizer

# Define the pattern
pattern = r"\b\w+(?:'\w+)?\b"

# Create a RegexpTokenizer with the pattern
tokenizer = RegexpTokenizer(pattern)

# Example tweet
tweet = "I'm excited about the iPhone's new features! Isn't it amazing?"

# Tokenize the tweet
tokens = tokenizer.tokenize(tweet)

# Print the tokens
print(tokens)


["I'm", 'excited', 'about', 'the', "iPhone's", 'new', 'features', "Isn't", 'it', 'amazing']


This output shows that the tokenizer breaks dwon the tweet into components,and handling even  shortened words and ownership accurately.

In [14]:
df1.sample(10)

tweet  \
4489  We all, including me, have this emotional thin...   
5438  RT @mention Apple is quarter of the music indu...   
5350  RT @mention 20 Percent of Google Searches are ...   
3357  CBS, SCVNGR Launch Spy Game at SXSW: If you ha...   
7403  #SXSW -Apple Pop-Up Store being set up right n...   
6544  RT @mention RT @mention Google to Launch Major...   
1097  We've got some exciting things that we'll be s...   
8857  I wish I was in Austin, if only for the free i...   
6400  RT @mention OH .. #vimeo: iPhone app submitted...   
1486  Very wise... RT @mention Apple is opening up a...   

                       target_entity                             emotion  \
4489                            iPad                    Positive emotion   
5438                           Apple                    Positive emotion   
5350                   Uncategorized  No emotion toward brand or product   
3357                   Uncategorized  No emotion toward brand or product   
7403                   Uncategorized  No emotion toward brand or product   
6544                   Uncategorized  No emotion toward brand or product   
1097                          iPhone                    Positive emotion   
8857                            iPad                    Positive emotion   
6400                   Uncategorized  No emotion toward brand or product   
1486  Other Apple product or service                    Positive emotion   

                                          cleaned_tweet  \
4489  we all including me have this emotional thing ...   
5438  rt mention apple is quarter of the music indus...   
5350  rt mention 20 percent of google searches are f...   
3357  cbs scvngr launch spy game at sxsw if you have...   
7403  sxsw apple popup store being set up right now ...   
6544  rt mention rt mention google to launch major n...   
1097  weve got some exciting things that well be sho...   
8857  i wish i was in austin if only for the free ip...   
6400  rt mention oh  vimeo iphone app submitted to s...   
1486  very wise rt mention apple is opening up a tem...   

                                       lemmatized_tweet  
4489  we all including me have this emotional thing ...  
5438  rt mention apple is quarter of the music indus...  
5350  rt mention 20 percent of google search are for...  
3357  cbs scvngr launch spy game at sxsw if you have...  
7403  sxsw apple popup store being set up right now ...  
6544  rt mention rt mention google to launch major n...  
1097  weve got some exciting thing that well be show...  
8857  i wish i wa in austin if only for the free ipa...  
6400  rt mention oh vimeo iphone app submitted to st...  
1486  very wise rt mention apple is opening up a tem...

Compared to other outputs this one right here differs from the rest becuase here we used sample and the samples of the rows we printed might containtweets with varying characteristicss.

In [15]:
df1['target_entity'].fillna('Uncategorized', inplace=True)